packages/web-forms/src/components/common/map/useMapBlock.ts

```diff
@@ -0,0 +1,282 @@
+import {
+	getSavedStyles,
+	getSelectedStyles,
+	getUnselectedStyles,
+} from '@/components/common/map/map-styles.ts';
+import type { FeatureCollection } from 'geojson';
+import { Map, MapBrowserEvent, View } from 'ol';
+import { Zoom } from 'ol/control';
+import Feature from 'ol/Feature';
+import GeoJSON from 'ol/format/GeoJSON';
+import { LineString, Point, Polygon } from 'ol/geom';
+import TileLayer from 'ol/layer/Tile';
+import WebGLVectorLayer from 'ol/layer/WebGLVector';
+import type { Pixel } from 'ol/pixel';
+import { fromLonLat } from 'ol/proj';
+import { OSM } from 'ol/source';
+import VectorSource from 'ol/source/Vector';
+import { computed, ref, shallowRef, watch } from 'vue';
+
+type GeometryType = LineString | Point | Polygon;
+
+const STATES = {
+	LOADING: 'loading',
+	READY: 'ready',
+	ERROR: 'error',
+} as const;
+
+const DEFAULT_GEOJSON_PROJECTION = 'EPSG:4326';
+const DEFAULT_VIEW_CENTER = [0, 0];
+const MAX_ZOOM = 16;
+const MIN_ZOOM = 2;
+const GEOLOCATION_TIMEOUT_MS = 10 * 1000;
+const ANIMATION_TIME = 1000;
+const SMALL_DEVICE_WIDTH = 576;
+const MAP_HIT_TOLERANCE = 5;
+const FEATURE_ID_PROPERTY = 'odk_feature_id';
+const SAVED_ID_PROPERTY = 'savedId';
+const SELECTED_ID_PROPERTY = 'selectedId';
+
+export function useMapBlock() {
+	const currentState = shallowRef<(typeof STATES)[keyof typeof STATES]>(STATES.LOADING);
+	const errorMessage = shallowRef<{ title: string; message: string } | undefined>();
+	let mapInstance: Map | undefined;
+	const savedFeature = ref<Feature<GeometryType> | undefined>();
+	const selectedFeature = ref<Feature<GeometryType> | undefined>();
+	const selectedFeatureProperties = computed(() => {
+		return selectedFeature.value?.getProperties();
+	});
+
+	const featuresSource = new VectorSource();
+	const featuresVectorLayer = new WebGLVectorLayer({
+		source: featuresSource,
+		style: [
+			...getUnselectedStyles(FEATURE_ID_PROPERTY, SELECTED_ID_PROPERTY, SAVED_ID_PROPERTY),
+			...getSelectedStyles(FEATURE_ID_PROPERTY, SELECTED_ID_PROPERTY, SAVED_ID_PROPERTY),
+			...getSavedStyles(FEATURE_ID_PROPERTY, SAVED_ID_PROPERTY),
+		],
+		variables: { [SAVED_ID_PROPERTY]: '', [SELECTED_ID_PROPERTY]: '' },
+	});
+
+	const initializeMap = (mapContainer: HTMLElement, geoJSON: FeatureCollection): void => {
+		if (mapInstance) {
+			return;
+		}
+
+		mapInstance = new Map({
+			target: mapContainer,
+			layers: [new TileLayer({ source: new OSM() }), featuresVectorLayer],
+			view: new View({ center: DEFAULT_VIEW_CENTER, zoom: MIN_ZOOM }),
+			controls: [new Zoom()],
+		});
+
+		currentState.value = STATES.READY;
+		loadGeometries(geoJSON);
+	};
+
+	const handleClick = (event: MapBrowserEvent) => selectFeatureByPosition(event.pixel);
+
+	const toggleClickBinding = (bindClick: boolean) => {
+		mapInstance?.un('singleclick', handleClick);
```

Comment:
I think @matthew-white commented somewhere but adding here for visibility... This feels sluggish and using `click` or similar might be better?